pilot/pkg/config/kube/crd/client.go

```diff
@@ -70,20 +60,81 @@ type IstioObjectList interface {
 
 // Client is a basic REST client for CRDs implementing config store
 type Client struct {
+	// Map of apiVersion to restClient.
+	clientset map[string]*restClient
+
+	// domainSuffix for the config metadata
+	domainSuffix string
+}
+
+type restClient struct {
+	apiVersion *schema.GroupVersion
+
+	// descriptor from the same apiVerion.
 	descriptor model.ConfigDescriptor
 
+	// types of the schema and objects in the descriptor.
+	types []*schemaType
+
 	// restconfig for REST type descriptors
 	restconfig *rest.Config
 
 	// dynamic REST client for accessing config CRDs
 	dynamic *rest.RESTClient
+}
 
-	// domainSuffix for the config metadata
-	domainSuffix string
+func apiVersion(schema *model.ProtoSchema) string {
+	return ResourceGroup(schema) + "/" + schema.Version
+}
+
+func apiVersionFromConfig(config *model.Config) string {
+	return config.Group + "/" + config.Version
+}
+
+func (cl *Client) newClientSet(descriptor model.ConfigDescriptor) error {
```

Comment:
I would expect `newClientSet` to create a new client, but here it just initializes a partially created object. This could be simplified by conslidated all of the client initialization code (lines 176-188 and `init` below) into `newClientSet` and having it return a new fully created client.

Reply:
I reorged newClientSet out of Client to make it more go style object creations. Does it feel better?

I want to avoid put everything into a large NewClient method.